### PR TITLE
Add view and edit capapabilities to sample page

### LINKED ISSUE #46: Implement individual sample page
When a user taps on an individual sample listed on a study page (https://github.com/microbiomedata/nmdc-field-notes/issues/30), they should be taken to a sample page. The mockups don't appear to spec out much in particular for this page. My general thought is that this could be a hybrid view/edit page. Since there are potentially a lot of metadata fields for each sample I'm not sure a monolithic form with an input for each field presented all at once will be optimal. 

Rather, I could envision a page that displays each field like a "view" page, but tapping on a field brings up an edit overlay for that single field. The overlay could also present descriptive information from the schema (see also https://github.com/microbiomedata/nmdc-server/issues/1166) similar to the help sidebar in the desktop submission portal. The main view should also give an indication of fields that don't validate. A very rough sketch:

Image

### PR DESCRIPTION
**Pull request by pkalita-lbl**
Fixes #46 

## Summary

This is a first pass at a single page to view and edit sample information with schema-based validation. The page initially shows a grouped list of sample slots (aka fields). The grouping and ordering is derived from the schema. Tapping on a slot brings up a modal overlay for editing the value of that slot. This overlay also provides help content derived from the schema and validation messages. 

### The `Validator` class

The `Validator` class (defined in `src/Validator.ts`) performs validation of objects based on a LinkML schema. This class and its test (`src/Validator.test.tsx`) were mostly lifted over from the `DataHarmonizer` package ([source](https://github.com/cidgoh/DataHarmonizer/blob/25a18fe64ea9c1c366a33ba5669b041a0054b45e/lib/Validator.js)) and stripped down a bit to account for the fact that 1) `DataHarmonizer` operates on data that comes in as a 2-D array of strings (i.e. spreadsheet-like) whereas this operates on arrays of objects and 2) `DataHarmonizer` has its own little validation quirks that we don't want to inherit. 

In a perfect world there would be a standalone NPM package for LinkML validation, and both this app and `DataHarmonizer` could rely on that package. I hope we can get there one day! But in the interest of time constraints, today is not that day.

A valid question is: why a custom validator? Why not use JSON Schema or some other off-the-shelf technology? LinkML is native format of the [submission schema](https://github.com/microbiomedata/submission-schema). JSON Schema can be derived from it (see? [here it is](https://github.com/microbiomedata/submission-schema/blob/main/project/jsonschema/nmdc_submission_schema.schema.json)), but you lose a bit in the process. For example you'd lose the comments, examples, and grouping/ranking information on slots. So we need the LinkML schema anyway for UI niceties, we might as well use it for validation too!

### The submission schema

The portal backend serves the currently-used submission schema at `/static/submission_schema/submission_schema.json`. The desktop submission portal picks up this file to drive its interface, so the app should as well! Then everything's nice and synchronized.

As noted, it's not a `/api` endpoint, but rather than introduce a new variable in `src/config.ts` I chose to strip the `/api` off of the existing one. This is also reflected in `.env.local.example` and `.env.staging`. Then I needed to add it back in places like `/src/mocks/server.ts`. The end result of all that shuffling is a new method in `src/api.ts` which fetches the schema.

In order to type the output of that new method (and to have type definitions in various other places that work with the schema) I've added the TypeScript definition of the [LinkML Metamodel](https://linkml.io/linkml/schemas/models.html#metamodel) in `src/linkml-metamodel.ts`. This file is _almost_ entirely auto-generated. See the comment at the top of the file for more information on that.

A new query (`useSubmissionSchema`) in `src/queries.ts` is how components can access the submission schema. It has a long `staleTime` because that file can only change with new versions of the portal backend. 

### The new page and components

The sample page (`src/pages/SamplePage/SamplePage.tsx`) was previously a stub. Now it responsible for:
* managing state like the submission schema, submission data, validation results, which slot (if any) is being edited
* performing simple validation as a user enters a new value into a slot (`handleModalValueChange`)
* performing full validation and saving to the server when a user finishes entering a new value into a slot (`handleSave` + `useEffect` callback)

It delegates the visual layout to two new components: `SampleView` and `SampleSlotEditModal`. 

`SampleView` simply groups the slots of a given schema class for display and then unrolls those groups of slots `IonItem`s separated by section headers.

`SampleSlotEditModal` is responsible for displaying a user input element along with validation and help content. The input element is either an `IonSelect` or `IonTextarea` depending on whether the range of the slot is an enum. There is also a bit of custom stuff in there to sneak in a clear button which `IonSelect` and `IonTextarea` don't provide (especially important in the `IonSelect` case -- we need to let the user "unchoose" an option!). 

This component also has some special custom behavior for certain slots. If it is displaying the `collection_date` slot, we add a button to populate the current date. If it is displaying the `lat_lon` slot, we add a button to populate using GPS. It might be nice to have a more generic way of doing this instead of just tacking on special cases inside of `SampleSlotEditModal`. It's probably fine for now, but other ideas are welcome.

### New dependencies

`@capacitor/geolocation` has been added to support populating location via GPS. I attempted to follow their [setup instructions](https://capacitorjs.com/docs/apis/geolocation#install), hence the updates to various `/ios` and `/android` files. I can't say I rigorously tested this on device simulators, but the web version works!

`date-fns` is used in a few places for translating between strings and `Date`objects during input and validation of slots whose range is a date or datetime.

### CODE REVIEW

**Comment by pkalita-lbl** [comment]
D'oh! After all that I still have failing tests. Will fix momentarily.

**Review by yxu-lanl — APPROVED** [review]

**Review by eecavanna — APPROVED** [review]
Looks great! Thanks for including that thorough PR description! 👏 

Things I did _not_ review:
- I did not review the entire `Validator.test.tsx` file, but the portion I did review looked good to me and I'm OK with it being merged in as is.
- I did not review `src/linkml-metamodel.ts` at all, except for the introductory comment. Given that it's a pretty much auto-generated file, I'm OK with it being merged in. I figure I'll learn about it over time.

I left some questions and comments.

In short, I am comfortable with the branch being merged in as is. I didn't come across anything that I would want to hold up a merge for.

File: src/Validator.test.tsx [hunk 1]
```diff
@@ -0,0 +1,1063 @@
...
+  },
+  time: {
+    name: "time",
+    uri: "xsd:time",
+  },
+};
+
+test("Validator should validate string types", () => {
+  const validator = new Validator({
+    id: "test",
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.test.tsx]: Haha there is a lot of "test"-ing going on. I pulled a lot of the test schema setup into a helper function to keep the actual test case functions a little more focused.

File: src/Validator.ts [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+    const slotMinimumValue = this.parseMinMaxConstraint(
+      slotDefinition.minimum_value,
+      slotType?.uri,
+    );
+    const slotMaximumValue = this.parseMinMaxConstraint(
+      slotDefinition.maximum_value,
+      slotType?.uri,
+    );
+
+    const validate = (value: unknown) => {
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.ts:165]: Yep your interpretation is correct. Added a comment.

File: src/Validator.test.tsx [hunk 3]
```diff
@@ -0,0 +1,1063 @@
...
+
+  let fn = validator.getValidatorForSlot("an_integer");
+  expect(fn(undefined)).toBeUndefined();
+  expect(fn(14)).toBeUndefined();
+  expect(fn(-14)).toBeUndefined();
+  expect(fn(3.1415)).toEqual("Value is not an integer");
+
+  fn = validator.getValidatorForSlot("a_float");
+  expect(fn(undefined)).toBeUndefined();
+  expect(fn(3.1415)).toBeUndefined();
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.test.tsx]: Done. Added additional test inputs.

File: src/Validator.test.tsx [hunk 4]
```diff
@@ -0,0 +1,1063 @@
...
+          },
+        },
+      },
+    },
+  });
+  validator.useTargetClass("Test");
+
+  const fn = validator.getValidatorForSlot("required_string");
+  expect(fn("correct")).toBeUndefined();
+  expect(fn("")).toEqual("This field is required");
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.test.tsx:199]: Added additional test cases here. Currently the validator _does not_ do any automatic `trim()`. There's a slippery slope of mixing automatic repair concerns and validation concerns. We should circle back on whether whitespace-only strings are being accepted for required slots, but if so I think the `trim()` should live outside the validator.

File: src/Validator.test.tsx [hunk 5]
```diff
@@ -0,0 +1,1063 @@
...
+  expect(fn(undefined)).toEqual("This field is required");
+});
+
+test("Validator should validate minimum and maximum numeric constraints", () => {
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.test.tsx]: It does support that! Added test cases for min/max floats.

File: src/Validator.test.tsx [hunk 6]
```diff
@@ -0,0 +1,1063 @@
...
+  expect(fn(["One", "Two", "Three"])).toBeUndefined();
+  expect(fn(["One", "whoops", "Three"])).toEqual("Value is not allowed");
+
+  fn = validator.getValidatorForSlot("many_small_integers");
+  expect(fn(undefined)).toBeUndefined();
+  expect(fn([1])).toBeUndefined();
+  expect(fn([1, 2, 3])).toBeUndefined();
+  expect(fn([1, 2.2, 3])).toEqual("Value is not an integer");
+  expect(fn([1, 2, "whoops"])).toEqual("Value is not an integer");
+  expect(fn([11, 2, 3])).toEqual("Value is greater than maximum value");
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.test.tsx:510]: Yes, currently that is the expected behavior. Added that test case.

File: src/Validator.ts [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+
+interface ValidatorOptions {
+  dateFormat?: string;
+  dateTimeFormat?: string;
+}
+
+export type ValidatorFn = (value: unknown) => string | undefined;
+export type ValidationResults = Record<number, Record<string, string>>;
+
+const REF_DATE = new Date(0);
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.ts:28]: Added

File: src/Validator.ts [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+    values.map((col) => col.length),
+  );
+
+  const previousValues = new Map();
+  for (let row = 0; row < longestColumn; row += 1) {
+    const rowValue = values.map((col) => String(col[row]));
+    // This should be a safe way to produce a unique hash for the row. If there
+    // are edge cases where this doesn't work we could look at more robust
+    // solutions (e.g. https://github.com/puleos/object-hash).
+    const rowHash = rowValue.join("\u0000");
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.ts:46]: The same logic is in DataHarmonizer and so far I haven't heard of any cases where it has been a problem.

File: src/Validator.ts [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+      slotDefinition.minimum_value,
+      slotType?.uri,
+    );
+    const slotMaximumValue = this.parseMinMaxConstraint(
+      slotDefinition.maximum_value,
+      slotType?.uri,
+    );
+
+    const validate = (value: unknown) => {
+      const valueDefined = value != null && value !== "";
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.ts:166]: Maybe this is a bad habit of mine but I tend to use `!=` or `==` when comparing with `null` because it implicitly also compares against `undefined`. So this:

```js
value != null
```

is equivalent to

```js
value !== null && value !== undefined
```

I feel somewhat justified in using that shortcut because ESLint carves out exceptions to their `eqeqeq` rule to allow such things: https://eslint.org/docs/latest/rules/eqeqeq.

Bigger picture, there are probably a lot places where we could tighten up our typing (e.g. to only allow undefined and not null), but I'm willing to save that for another time.

File: src/Validator.ts [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+            (slotDefinition.equals_number !== undefined &&
+              value !== slotDefinition.equals_number)
+          ) {
+            return "Value does not match constant";
+          }
+
+          if (
+            slotDefinition.pattern !== undefined &&
+            typeof value === "string" &&
+            !value.match(slotDefinition.pattern)
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/Validator.ts:264]: LinkML doesn't explicitly state it (see also https://github.com/linkml/linkml/issues/1695). It's just kind of determined by whatever environment the schema or an artifact generated from it is going into. So the general recommendation is "don't use anything too exotic in your patterns".

File: src/components/SampleSlotEditModal/SampleSlotEditModal.tsx [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+              <IonLabel color="warning">{validationResult}</IonLabel>
+            </IonItem>
+          )}
+
+          {slot.name === "collection_date" && (
+            <IonButton
+              className="ion-padding-vertical"
+              expand="block"
+              onClick={() =>
+                handleValueChange(format(new Date(), "yyyy-MM-dd"))
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/components/SampleSlotEditModal/SampleSlotEditModal.tsx:150]: No it's ambiguous -- not sure if that was an intentional decision or not (probably not).

File: src/components/SchemaSlotHelp/SchemaSlotHelp.tsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    } else if (hasMinValue) {
+      paragraph += `greater than or equal to ${slot.minimum_value}.`;
+    } else if (hasMaxValue) {
+      paragraph += `less than or equal to ${slot.maximum_value}.`;
+    }
+    guidanceParagraphs.push(paragraph);
+  }
+  if (slot.identifier) {
+    guidanceParagraphs.push(
+      "Each record must have a unique value for this field.",
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/components/SchemaSlotHelp/SchemaSlotHelp.tsx]: Theoretically we could use this component for slots that aren't attached to samples. Who knows some day maybe we will! So I've added a `containingObjectName` prop to this component and when we use it in `SampleSlotEditModal` I've passed the value `"sample"`.

File: src/linkml-metamodel.ts [hunk 13]
```diff
@@ -0,0 +1,896 @@
+/* This file is /mostly/ autogenerated from the LinkML Metamodel.
+ * To regenerate this file from a python project with the `linkml`
+ * package installed:
+ *
+ *   > gen-typescript .venv/lib/python3.9/site-packages/linkml_runtime/linkml_model/model/schema/meta.yaml > linkml-metamodel.ts
+ *
+ * Since the metamodel does not change too frequently, we don't need
+ * to worry about always keeping it updated.
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/linkml-metamodel.ts:8]: The metamodel source is here: https://github.com/linkml/linkml-model. In practice a metamodel release don't have any practical impact until it is brought into https://github.com/linkml/linkml-runtime. Most `linkml-runtime` releases _don't_ change the metamodel and when they do they usually increment the minor version number.

File: src/utils.ts [hunk 14]
```diff
@@ -1,11 +1,94 @@
...
+  slots: SlotDefinition[];
+}
+export function groupClassSlots(
+  schemaDefinition: SchemaDefinition,
+  className: string,
+): SlotGroup[] {
+  const classDefinition = schemaDefinition.classes?.[className];
+  if (!classDefinition) {
+    throw new Error(`Class ${className} not found in schema`);
+  }
```
[the remark this answers is not in the document]
Comment (pkalita-lbl) [reply src/utils.ts:51]: You could pass this function a `SchemaDefinition` that includes classes `A`, `B`, and `C`, but ask the function to group the slots of class `D`. That would cause the lookup in the previous line to return `undefined`. Throwing an error in that case is maybe a bit extreme. If it turns out to be a problem we could also `return []` at that point.
